### PR TITLE
[ENH] VMD (variational mode decomposition) transformer based on `vmdpy`

### LINKED ISSUE #5128: [BUG] - TransformedTargetForecaster with Arima can't fit a transformer taking series as input and returning pd.Dataframe with multiple series
**Describe the bug**

When using a custom decomposition transformer which takes a single time series input and produces multiple in a Pandas dataframe format, used in conjunction with the TransformedTargetForecaster as a pipeline and the Arima model as the forecaster, will produce the following error when making predictions:
    raise ValueError("input must be univariate pd.DataFrame, with one column")
ValueError: input must be univariate pd.DataFrame, with one column

**To Reproduce**

```python
import pandas as pd
import numpy as np
from sktime.forecasting.arima import ARIMA
from sktime.forecasting.compose import TransformedTargetForecaster
from sktime.forecasting.model_evaluation import evaluate
from sktime.forecasting.model_selection import ExpandingWindowSplitter
from sktime.performance_metrics.forecasting import MeanSquaredError

from vmd_specifc.VmdTransformer import VmdTransformer
#. Time Domain 0 to T
T = 1000
fs = 1/T
t = np.arange(1,T+1)/T
freqs = 2*np.pi*(t-0.5-fs)/(fs)

#. center frequencies of components
f_1 = 2
f_2 = 24
f_3 = 288

#. modes
v_1 = (np.cos(2*np.pi*f_1*t))
v_2 = 1/4*(np.cos(2*np.pi*f_2*t))
v_3 = 1/16*(np.cos(2*np.pi*f_3*t))

f = v_1 + v_2 + v_3 + 0.1*np.random.randn(v_1.size)
data = pd.Series(f)

pipe = TransformedTargetForecaster(steps=[
    ("vmd transformer", VmdTransformer()),
    ("forecaster", ARIMA())
])
# the below commented section also show the same problem
# y = data[:(round((len(data) / 100) * 70))]
# pipe.fit(y)
# preds = pipe.predict(fh=[1,2,3,4,5,6])
cv = ExpandingWindowSplitter(step_length=2, fh=2,
                             initial_window=(round((len(data) / 100) * 70)) )
cv_df = evaluate(error_score='raise', forecaster=pipe, y=data, cv=cv, strategy="refit",
                 return_data=True,
                 scoring=MeanSquaredError(square_root=True))

```

**Expected behavior**

A separate Arima model should be fit on each column of the returned pandas data frame, and the model should not throw an error.
**Additional context**

The link to the custom VmdTransformer pull request will be given here when created, or mentioned in the comments.
**Versions**
<details>

System:
    python: 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)]
executable: C:\Users\Dane\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\python.exe
   machine: Windows-10-10.0.19045-SP0

Python dependencies:
          pip: 23.0
       sktime: 0.21.0
      sklearn: 1.2.1
       skbase: 0.4.6
        numpy: 1.22.4
        scipy: 1.10.0
       pandas: 1.5.3
   matplotlib: 3.3.4
       joblib: 1.2.0
  statsmodels: 0.13.2
        numba: 0.55.2
     pmdarima: 2.0.2
      tsfresh: 0.20.0
   tensorflow: None
tensorflow_probability: 0.19.0
</details>

### PR DESCRIPTION
**Pull request by DaneLyttinen**
#### Reference Issues/PRs

Implements #5137 and fixes #5128

#### What does this implement/fix? Explain your changes.

This implementation introduces a new transformer, a Variational mode decomposition transformer of which takes an original time series and transforms it into multiple Intrinsic mode Functions, the number of modes creates is specified by the user-supplied K parameter. If the user does not supply such a parameter, the transformer will find the number of modes dynamically with some parameters affecting how the parameters will be found.

Depends on https://github.com/sktime/sktime/pull/5166 for the `inverse_transform`-not-exact tag.

#### Does your contribution introduce a new dependency? If yes, which one?

No, `vmdpy` is moved into `sktime`.

#### What should a reviewer concentrate their feedback on?

This is still in development, currently, it does not fit with a TransformedTargetForecaster for example and this pull request was created for the bug mentioned in #5128. There are also some caveats which need to be considered, 
for example: 
* The package truncates the last row if the resulting series is an odd number therefore, currently it just appends the last row to make the time series the same, but this might not be what users would like. Is there another way to handle this so the truncation is allowed rather than trying to append the last row again to make the length the same size?

#### Did you add any tests for the change?

Not currently no.

#### Any other comments?

Please let me know of any required changes to make the transformer more usable :) 
##### For all contributions
- [] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/transformations/series/vmd.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    init : int, optional (default=1)
+        parameter for omegas, default of one will initialize the omegas uniformly
+    tol : int, optional (default=1e-7)
+        parameter for omegas, default of one will initialize the omegas uniformly
+
+    """
+
+    _tags = {
+        "scitype:transform-input": "Series",
+        "scitype:transform-output": "Primitives",
```
Comment (fkiraly) [inline sktime/transformations/series/vmd.py:49]: output should be `"Series"`, is a decomposition into mutliple time series

File: sktime/transformations/series/vmd.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        "scitype:instancewise": True,
+        "scitype:transform-labels": "None",
+        "X_inner_mtype": "pd.DataFrame",
+        "y_inner_mtype": "None",
+        "univariate-only": False,
```
Comment (fkiraly) [inline sktime/transformations/series/vmd.py:54]: Does `VMD` take multiple columns, i.e., an arbitrary 2D array?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

This is very useful, using this we should be able to make it work!

Saying ahead, I'm unfamiliar with the `vmdpy` package, so you may have to help me here, @DaneLyttinen.
Below are some points to ensure compatibility with `sktime` contracts.

* above some comments regarding the tags - if I understand correctly, VMD decomposes a time series into multiple time series. So the output type should be `"Series"`. From the code, I think this also works only for one column/variable? So the `univariate-only` tag should be `True`.
* the `_inverse_transform` should ensure that the column name is the same as seen in `_fit`.
* the soft dependency `pyvmd` should be imported only inside the class - I assume, in `__runVMDetc` and in `_transform`.
* for the `sktime` tests to work, the package `pyvmd` should be added in `pyproject.toml`, in the dependency set `all_extras`, and/or `all_extras_pandas1` (depending on `pandas` version compatibility)

**Comment by DaneLyttinen** [comment]
Cool, 

I've made the changes as suggested, although the issue still persists, unfortunately.
The package is quite tiny, it's essentially just a file with ~140 lines but just don't want to reinvent the wheel when a solution is already out there :)

**Comment by fkiraly** [comment]
> The package is quite tiny, it's essentially just a file with ~140 lines but just don't want to reinvent the wheel when a solution is already out there :)

Makes sense, as long as it is well-maintained?

I'm somewhat concerned about it being 3 years since the last release, it may be abandonend.

For this PR, we can use it as an import - if we want to add the estimator to `sktime` at the end, we could cosider copy-pasting, as that could be better security- and maintenance-wise. With permissio nof the author, and they would be credited either way.

As said, for now, let's import until we have diagosed your issue.

**Comment by fkiraly** [comment]
the CI is failing with code formatting issues - would e great if you could fix while I look at the internals

**Comment by fkiraly** [comment]
@DaneLyttinen, would you be so kind and get the file to a state where:

* code formatting checks pass
* the basic contact tests with `check_estimator` pass?

You can run the contract tests via

```python
from sktime.transformations.series.vmd import VmdTransformer
from sktime.utils.estimator_checks import check_estimator

check_estimator(VmdTransformer, raise_exceptions=True)
```

I've already fixed a couple, so I suspect you weren't following the extension guide here?
https://www.sktime.net/en/latest/developer_guide/add_estimators.html

What I fixed:
* `_fit` should not change `__init__` parameters, but it changed `self.K`
* `_transform` should not write to `self`, but it wrote `self.fit_column_names`

Now, these don't necessarily relate to the issue you reported, but we need to start with a contract compliant transformer.

I think I've fixed everything except code formatting how, but I'm getting a lot of warnings, this might indicate some issue with interfacing?
Here's a sample:

```
[c:\ProgramData\Anaconda3\envs\sktime-skbase-311\Lib\site-packages\vmdpy\vmdpy.py:114](file:///C:/ProgramData/Anaconda3/envs/sktime-skbase-311/Lib/site-packages/vmdpy/vmdpy.py:114): RuntimeWarning: invalid value encountered in scalar divide
  omega_plus[n+1,k] = np.dot(freqs[T//2:T],(abs(u_hat_plus[n+1, T//2:T, k])**2))/np.sum(abs(u_hat_plus[n+1,T//2:T,k])**2)
[c:\ProgramData\Anaconda3\envs\sktime-skbase-311\Lib\site-packages\vmdpy\vmdpy.py:112](file:///C:/ProgramData/Anaconda3/envs/sktime-skbase-311/Lib/site-packages/vmdpy/vmdpy.py:112): RuntimeWarning: invalid value encountered in divide
  u_hat_plus[n+1,:,k] = (f_hat_plus - sum_uk - lambda_hat[n,:]/2)/(1+Alpha[k]*(freqs - omega_plus[n,k])**2)
[c:\ProgramData\Anaconda3\envs\sktime-skbase-311\Lib\site-packages\vmdpy\vmdpy.py:114](file:///C:/ProgramData/Anaconda3/envs/sktime-skbase-311/Lib/site-packages/vmdpy/vmdpy.py:114): RuntimeWarning: invalid value encountered in scalar divide
  omega_plus[n+1,k] = np.dot(freqs[T//2:T],(abs(u_hat_plus[n+1, T//2:T, k])**2))/np.sum(abs(u_hat_plus[n+1,T//2:T,k])**2)
[c:\ProgramData\Anaconda3\envs\sktime-skbase-311\Lib\site-packages\vmdpy\vmdpy.py:112](file:///C:/ProgramData/Anaconda3/envs/sktime-skbase-311/Lib/site-packages/vmdpy/vmdpy.py:112): RuntimeWarning: invalid value encountered in divide
  u_hat_plus[n+1,:,k] = (f_hat_plus - sum_uk - lambda_hat[n,:]/2)/(1+Alpha[k]*(freqs - omega_plus[n,k])**2)
[c:\ProgramData\Anaconda3\envs\sktime-skbase-311\Lib\site-packages\vmdpy\vmdpy.py:114](file:///C:/ProgramData/Anaconda3/envs/sktime-skbase-311/Lib/site-packages/vmdpy/vmdpy.py:114): RuntimeWarning: invalid value encountered in scalar divide
  omega_plus[n+1,k] = np.dot(freqs[T//2:T],(abs(u_hat_plus[n+1, T//2:T, k])**2))/np.sum(abs(u_hat_plus[n+1,T//2:T,k])**2)
[c:\ProgramData\Anaconda3\envs\sktime-skbase-311\Lib\site-packages\vmdpy\vmdpy.py:112](file:///C:/ProgramData/Anaconda3/envs/sktime-skbase-311/Lib/site-packages/vmdpy/vmdpy.py:112): RuntimeWarning: invalid value encountered in divide
  u_hat_plus[n+1,:,k] = (f_hat_plus - sum_uk - lambda_hat[n,:]/2)/(1+Alpha[k]*(freqs - omega_plus[n,k])**2)
[c:\ProgramData\Anaconda3\envs\sktime-skbase-311\Lib\site-packages\vmdpy\vmdpy.py:114](file:///C:/ProgramData/Anaconda3/envs/sktime-skbase-311/Lib/site-packages/vmdpy/vmdpy.py:114): RuntimeWarning: invalid value encountered in scalar divide
  omega_plus[n+1,k] = np.dot(freqs[T//2:T],(abs(u_hat_plus[n+1, T//2:T, k])**2))/np.sum(abs(u_hat_plus[n+1,T//2:T,k])**2)
[c:\ProgramData\Anaconda3\envs\sktime-skbase-311\Lib\site-packages\vmdpy\vmdpy.py:112](file:///C:/ProgramData/Anaconda3/envs/sktime-skbase-311/Lib/site-packages/vmdpy/vmdpy.py:112): RuntimeWarning: invalid value encountered in divide
  u_hat_plus[n+1,:,k] = (f_hat_plus - sum_uk - lambda_hat[n,:]/2)/(1+Alpha[k]*(freqs - omega_plus[n,k])**2)
```

**Comment by fkiraly** [comment]
PS: just in case you didn't notice (and I hope you don't mind), I pushed my fixes to the contract non-compliance issues on this branch. Not sure whether that's the best way to fix them, I made some educated guesses - feel free to change.

**Comment by fkiraly** [comment]
I think I found the problem.

There were both issues in your implementation around `inverse_transform`, and an actual bug (that is related but did not impact the problem you were facing).

* regarding `inverse_transform`, you need it to invert `transform`, so assume an output of the shape and format like from `transform`.
* further, the `skip-inverse-transform` tag needs to be `False`, this tells pipelines like `TransformedTargetForecaster` to skip it or not

Now, there also was a bug here - with vectorization of `inverse_transform`. It's a curious interaction, and not sure how to fix it, appearing when `univariate-only` is `True`, and `transform` produces multiple columns. In such a case, the `inverse_transform` does not know which columns to invert.

The solution is probably going to be a bit involved, I'll see what we can do.

It should however not impact your use case - you code in https://github.com/sktime/sktime/issues/5128 should now work with the modifications I've made (fixing `inverse_transform`)

I would highly appreciate if, in return, you could fix the remaining code formatting issues in this PR.

**Comment by fkiraly** [comment]
@DaneLyttinen, regarding the more general maintenance issue, I've opened an issue here: https://github.com/sktime/sktime/issues/5137

It seems the original `vmdpy` isn't robustly maintained, bugfixes and improvements seem to happen in forks which is not a sign of health.

**Comment by DaneLyttinen** [comment]
Fantastic, thank you very much for the assistance @fkiraly.

I must have missed that documentation to follow, apologies for that, I will fix the code formatting issues ASAP.

And yeah, it makes sense, although as a general package, I don't think anything would change regarding the implementation of VMD (I believe the Python implementation is ported from Matlab, which also seems to be dead), but it makes sense to port it to Sktime as well as making it more conformant and or performant. There is increasing research interest in using this as a method for time series predictions. Could possibly create it from scratch, I could (possibly) have an attempt at this although don't want to set any hard promises.

**Comment by DaneLyttinen** [comment]
I am going to add an example in the docstring soon. I didn't find a module to put the transformer under in the transformations.rst file, maybe wavelets? Or could make a new one under decomposition or the term used in her which seems to be reduction

**Comment by fkiraly** [comment]
> . I didn't find a module to put the transformer under in the transformations.rst file, maybe wavelets? Or could make a new one under decomposition or the term used in her which seems to be reduction

I would put it under the current section "detrending" which you could rename to "detrending and decomposition".

There are already decomposition methods in there, the `Deseasonalizer` and `STLTransformer`.

**Comment by fkiraly** [comment]
> I must have missed that documentation to follow, apologies for that, I will fix the code formatting issues ASAP.

Thanks!
There's still a couple, see above.

**Comment by fkiraly** [comment]
> Could possibly create it from scratch, I could (possibly) have an attempt at this although don't want to set any hard promises.

Sounds exciting and would be appreciated.
For planning this, the best place would be the issue #5137, maybe someone wants to help!

**Comment by fkiraly** [comment]
PS: I've run the `check_estimator` tests locally, and it complains with `transform` and `inverse_transform` not being inverse to each other. That would be odd, at least per my methdologial understanding, this should be a full additive decomposition?

**Comment by DaneLyttinen** [comment]
Will have a look at the new code formatting.

So VMD is mainly used for signals where the assumption is the number of frequency components is known beforehand, then it should create modes that represent those frequency oscillations (either exactly or in the least squares sense) that sum to the original signal. 

However, if this is not known, which it isn't in the test cases, the modes created will try and find them dynamically through the energy loss coefficient while this suffers from the underbinning and overbinning problem. The implementation in the paper here suggests a way to overcome this problem but personally got stuck trying to implement this due to the vague definition of counting 'extreme points' in the frequency domain.

https://www.sciencedirect.com/science/article/pii/S088832701830027X

So, in all, there will be some information loss when performing the inverse transform but in most cases, this will be very minimal, especially for longer time series and will suit the use cases that do not require 100% exact reconstruction.

**Comment by fkiraly** [comment]
> Will have a look at the new code formatting.

Thanks.

>So, in all, there will be some information loss when performing the inverse transform but in most cases,

Hm, then we need to modify the contract (and tests) to allow non-exact inverse transform. So far it was always assumed that if the inverse transform is implemented, it is an exact inverse.

**Comment by fkiraly** [comment]
> Will have a look at the new code formatting.

PS: if you're not sure how to get this all fixed, or automatically setup locally - feel free to attend one of the regular meet-ups for a screenshare! It might be fiddly for a first time setup, it's fiddly for everyone...

**Comment by fkiraly** [comment]
FYI, @DaneLyttinen, @vcarvo, following discussion in #5137, I've modified this PR to pull `vmdpy` directly into `sktime` where it can be maintained and updated.

**Review by fkiraly — APPROVED** [review]
I've fixed the outstanding issue - linting - while moving `vmdpy` into `sktime`, so this should now be ready to go.

**Review by fkiraly — DISMISSED** [review]
Approve original changes - my changes are only linting and moving `vmdpy` pkg in for future maintenance.

**Comment by DaneLyttinen** [comment]
Thanks @fkiraly, My apologies for being quite late, had to juggle quite a few things recently. 

I will add the transformer to the trasnformations.rst file and add an example to make it more complete as well.

**Comment by fkiraly** [comment]
oops, looks like you introduced some unintentional linting problems, @DaneLyttinen - you should really set this up locally :-)

Seems like an easy one to fix, let's make this a fun scavenger hunt. Hint:
image

**Review by fkiraly — DISMISSED** [review]
Fixed it in order to resolve merge conflicts with the MSTL transformer (which is also a decomposition, uses the same tag and has the same API ref section)

**Pull request merged**